### ISSUE TITLE
Allow users to opt-out of tracking errors notifications

### ISSUE THREAD

**Issue by tegioz** [artifacthub/hub#589]

**Comment by mattfarina** [comment]
Can we have the case where we are notified once or once every X weeks? Not just an opt-out but a rate limiting of notifications.

An idea would be to hash the error. If the error has not changed you don't need to re-notify.

**Comment by tegioz** [comment]
We've already done that @mattfarina , if the error has not changed the email shouldn't be sent again. So we shouldn't be bothering users often. 

The idea of this was to not bother them at all if they don't want to. But honestly, I think it'd be great if they always receive them given that they won't be that many and they may be highlighting an important issue. So I'm happy to close this if you don't think it's really necessary.

**Comment by mattfarina** [comment]
@tegioz I'll keep an eye on it. I've gotten the same error more than once today.

**Comment by tegioz** [comment]
Maybe it got cleared and then it came back again. I've seen that happening locally, specially when forcing all packages to be reindexed. 

Some http servers don't like us hitting them so fast and reject some requests. When that happens, errors can also be slightly different. I've had to add some rate limiting already for some cases (getting Charts packages from Github, for example, or we start getting 429 errors), maybe some more need it too.

Please let me know if you receive the same email (exactly the same errors) twice in a row within 30 mins. The trackers run at 0 and 30 every hour, and the same email should not be sent twice in a row. If you got two of them by an hour or more of difference it'd mean that the error got cleared (or changed) in one of the runs and it showed up again.

**Comment by tegioz** [comment]
I've taken a quick look at the database to check how the notifications were going @mattfarina. 7 out of the 102 repos we have registered have some errors.

This is the number of notifications sent for each of them. Except for Helm Stable and Incubator, we've only generated one notification for each of the repos during the last 5 hours (~10 tracker executions), which is good news.

```
 helm-stable         |      5
 helm-incubator      |      5
 billimek-charts     |      1
 armory              |      1
 routr               |      1
 community-operators |      1
 as                  |      1
```

For Helm Stable, we see one per hour, which suggests the errors are getting cleared and coming back on the next execution, as they happen every 30 mins. 

```
 helm-stable | 2020-08-18 18:00:25.892704+00
 helm-stable | 2020-08-18 17:00:25.737459+00
 helm-stable | 2020-08-18 16:00:25.519786+00
 helm-stable | 2020-08-18 15:00:25.348155+00
 helm-stable | 2020-08-18 14:00:25.151506+00
```

For Helm Incubator, we see kind of the same except for the executions at 16:00 and 16:30. I suspect the errors were different in those two notifications.

```
 helm-incubator | 2020-08-18 18:00:25.886364+00
 helm-incubator | 2020-08-18 16:30:25.664399+00
 helm-incubator | 2020-08-18 16:00:25.553496+00
 helm-incubator | 2020-08-18 15:00:25.360977+00
 helm-incubator | 2020-08-18 14:00:25.142478+00
```

If I remember correctly from when I've seen them locally, the errors in Stable and Incubator repos were happening while fetching some images, but only sometimes. Please let me know if I can be of any help.

**Comment by scottrigby** [comment]
I initially made this as a separate issue, but from Slack chat with @tegioz it's better to keep as a comment in this issue.

### feat: Enable/disable email notifications per repo

GitHub for example allows this per repo: [Managing security and analysis settings for your repository](https://docs.github.com/en/github/administering-a-repository/managing-security-and-analysis-settings-for-your-repository).

This is for the use case where a user is a member of an Artifact Hub Organization with multiple Repositories, where the user themselves is only involved in maintaining a subset of the Repositories and so can't do anything to stop the notifications. However would still like notifications for Repositories they are able to fix.

**Comment by tegioz** [comment]
Hi @mattfarina @scottrigby 

I've been looking this morning into the particular problem with the `Helm stable and incubator` repositories. The goal was to find out why did you keep receiving those emails when you really shouldn't (when they are enabled, as they are disabled at the moment). 

Some context first:

- Most if not all the errors are related to logos images (404 errors)
- All of the ones I've tried manually are legit errors, those urls don't work anymore
- Every time a repo is processed, a set of errors is generated and sorted (if there are any errors)
- Only when those errors change from execution to execution an event is generated and notifications may be triggered

But let's take a step back first (some more context):

- Errors may be generated when packages are processed by the tracker
- The Helm tracker, on a given repo, during the first execution processes all available packages
- On subsequent executions, it only processes new available packages and/or new versions, as well as **the ones that have changed**
- This means that, if the first processing of a repo generates some errors, the following executions most likely will not (unless a new package or version is released before it happens) and errors will be cleared
- There is a way to force all of them to be processed again, even if they haven't changed, which I usually use to index new metadata (not used often and not relevant for this case)

So let's say a package called *package1* with only one version available *1.0.0* was one of the ones triggering some errors. Once it's processed, the error shouldn't have never been triggered again as that package wasn't supposed to be processed again until a new version comes out. **Unless the package latest version has changed**. A package is considered to be changed when *its digest changes*. And I think that is exactly what's going on. IIRC we discussed something similar back in April @mattfarina, when the same packages continued appearing in the home page as recently updated packages even though the a new version hadn't been released.

To summarize, somehow Charts versions digest in the `index.yaml` file of the stable and incubator repos is changing every now and then (sometimes every hour). Not all of them all the time, just some. Because of this, the Artifact Hub Helm tracker is reindexing some packages as they are supposed to have changed. This can generate some errors, and given the changed packages are not always the same, the errors can be different, which would trigger the corresponding events and notifications.

It looks like these changes in Charts digests when they happen, happen with a minimum interval of 1 hour. The tracker runs every 30 mins, so the errors will be picked on one of the two executions within the hour, and cleared on the next one. The sample timeline below may illustrate the issue with more clarity:

```
23:00 | some charts digest updated
00:01 | tracker run 1 (errors generated)
00:31 | tracker run 2 (errors cleared as no new or updated packages were processed since last execution)
01:00 | some charts digest updated
01:01 | tracker run 3 (errors generated again, maybe the same ones or others, not relevant)
01:31 | tracker run 4 (errors cleared as no new or updated packages were processed since last execution)
...
```

To prove this, I've downloaded the Helm stable repo `index.yaml` file twice during the morning with one hour of difference between them.

### Some excerpts from sample 1:

```yaml
  airflow:
  - apiVersion: v1
    appVersion: 1.10.10
    created: 2020-08-25T09:29:22.297760129Z
    description: Airflow is a platform to programmatically author, schedule and monitor
      workflows
    digest: 618126f3d9e994d50ebfb4da974de8270a1736ae06b45db8709aa451544e6a6a
    home: https://airflow.apache.org/
    icon: https://airflow.apache.org/_images/pin_large.png
    keywords:
    - workflow
    - dag
    name: airflow
    sources:
    - https://airflow.apache.org/
    urls:
    - https://kubernetes-charts.storage.googleapis.com/airflow-7.5.0.tgz
    version: 7.5.0
```

```yaml
  elastic-stack:
  - apiVersion: v1
    appVersion: "6"
    created: 2020-08-25T09:29:22.342662538Z
    description: A Helm chart for ELK
    digest: c6b00e494497cffda71aa3fad2e2f4d58a5713df771f458e3d5c28d04235c94c
    home: https://www.elastic.co/products
    name: elastic-stack
    urls:
    - https://kubernetes-charts.storage.googleapis.com/elastic-stack-2.0.3.tgz
    version: 2.0.3
```

### Some excerpts from sample 2:

```yaml
  airflow:
  - apiVersion: v1
    appVersion: 1.10.10
    created: 2020-08-25T10:28:38.830572455Z
    description: Airflow is a platform to programmatically author, schedule and monitor
      workflows
    digest: d19cae7e4e73dacffe609f0190f45b1a93c955317412e011fec6d105c2da23ed
    home: https://airflow.apache.org/
    icon: https://airflow.apache.org/_images/pin_large.png
    keywords:
    - workflow
    - dag
    name: airflow
    sources:
    - https://airflow.apache.org/
    urls:
    - https://kubernetes-charts.storage.googleapis.com/airflow-7.5.0.tgz
    version: 7.5.0
```

```yaml
  elastic-stack:
  - apiVersion: v1
    appVersion: "6"
    created: 2020-08-25T10:28:38.871073656Z
    description: A Helm chart for ELK
    digest: bf72c28d0929ed30813b0b0638a19e05f983557a51bc91350d3f513f6e8a67ce
    home: https://www.elastic.co/products
    name: elastic-stack
    urls:
    - https://kubernetes-charts.storage.googleapis.com/elastic-stack-2.0.3.tgz
    version: 2.0.3
```

Please note that, even though the version is the same, the digest and the created timestamp have changed. And this is causing the tracker to reindex some existing versions even though it looks like they haven't really changed.

The fact that the created timestamp changes constantly is another issue by itself, as it is used by the Hub and displayed in the UI. It's also a problem if we'd like to highlight packages recently updated as we used to do on initial versions, as those which are being randomly updated constantly would win over real packages updates.

There are more differences between those two, and the Charts affected are not always the same if you keep comparing them over time. Those were just examples, a quick diff will show all of them.

Hope this helps solving the issue, please let me know your thoughts :)

**Comment by scottrigby** [comment]
That explains repeat alerts. Now, I wonder why is the digest changing… 🤔 💭
